resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/tpr/TPRCrudCalls.scala

```diff
@@ -0,0 +1,262 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.deploy.kubernetes.tpr
+
+import java.util.concurrent.{ThreadPoolExecutor, TimeUnit}
+
+import io.fabric8.kubernetes.client.{BaseClient, KubernetesClient}
+import okhttp3.{MediaType, OkHttpClient, Request, RequestBody, Response}
+import okio.{Buffer, BufferedSource}
+import org.json4s.{DefaultFormats, Formats}
+import org.json4s.JsonDSL._
+import org.json4s.jackson.JsonMethods.{compact, parse, pretty, render}
+import org.json4s.jackson.Serialization.{read, write}
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.{blocking, Future, Promise}
+import scala.util.control.Breaks.{break, breakable}
+
+import org.apache.spark.deploy.kubernetes.ClientBuilder
+import org.apache.spark.deploy.kubernetes.constants._
+import org.apache.spark.internal.Logging
+import org.apache.spark.SparkException
+import org.apache.spark.util.ThreadUtils
+
+/**
+ * Isolated this since the method is called on the client machine
+ */
+
+private[spark] case class Metadata(name: String,
+                                   uid: Option[String] = None,
+                                   labels: Option[Map[String, String]] = None,
+                                   annotations: Option[Map[String, String]] = None)
+
+private[spark] case class SparkJobState(apiVersion: String,
+                                        kind: String,
+                                        metadata: Metadata,
+                                        spec: Map[String, Any])
+
+private[spark] case class WatchObject(`type`: String, `object`: SparkJobState)
+
+private[spark] abstract class TPRCrudCalls extends Logging {
+  protected val k8sClient: KubernetesClient
+  protected val kubeMaster: String = KUBERNETES_MASTER_INTERNAL_URL
+  protected val kubeToken: Option[String] = None
+
+  implicit val formats: Formats = DefaultFormats + JobStateSerDe
+
+
+  protected val httpClient: OkHttpClient = ClientBuilder
+    .buildOkhttpClientFromWithinPod(k8sClient.asInstanceOf[BaseClient])
+
+  protected val namespace: String = k8sClient.getNamespace
+  private var watchSource: BufferedSource = _
+  private lazy val buffer = new Buffer()
+  protected implicit val ec: ThreadPoolExecutor = ThreadUtils
+    .newDaemonCachedThreadPool("tpr-watcher-pool")
+
+  private def executeBlocking(cb: => WatchObject): Future[WatchObject] = {
```

Comment:
Thanks Matt for taking a look at this. This is mostly a port of work we did a couple months ago. At the time, we considered making a contribution upstream, but we decided to go the crud way to save some cycles. 

I'll spend sometime looking into what it'll take to get a contribution upstream and maybe we could also talk about this at the next sig.